independent-projects/extension-maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java

```diff
@@ -678,17 +679,56 @@ public boolean visitLeave(DependencyNode node) {
     }
 
     private void addSource(ObjectNode extObject) throws MojoExecutionException {
-        Map<String, String> repo = ScmInfoProvider.getSourceRepo();
+        Scm scm = getScm();
+        String scmUrl = scm != null ? scm.getUrl() : null;
+
+        Map<String, String> repo = new ScmInfoProvider().getSourceRepo(scmUrl);
         if (repo != null) {
             ObjectNode metadata = getMetadataNode(extObject);
             for (Map.Entry<String, String> e : repo.entrySet()) {
-                // Tools may not be able to handle nesting in metadata, so do fake-nesting
-                metadata.put("scm-" + e.getKey(), e.getValue());
+                // Ignore if already set
+                String value = e.getValue();
+                String fieldName = "scm-" + e.getKey();
+                if (!metadata.has(fieldName) && value != null) {
+                    // Tools may not be able to handle nesting in metadata, so do fake-nesting
+                    metadata.put(fieldName, value);
+                }
 
             }
         }
     }
 
+    private Scm getScm() {
+        // We have three ways to do this; project.getScm() will query the derived model. Sadly, inherited <scm> entries are usually wrong, unless the parent is in the same project
+        // We can use getOriginalModel and getParent to walk the tree, but this will miss parents in poms outside the current execution
+        //We can also use more sophisticated techniques using repository managers, but we need to handle errors
+
+        Scm scm = null;
+        try {
+            LocalProject localProject = new BootstrapMavenContext(BootstrapMavenContext.config()
+                    .setCurrentProject(project.getBasedir().toString())
+                    .setRemoteRepositoryManager(remoteRepoManager)
+                    .setRemoteRepositories(repos)
+                    .setEffectiveModelBuilder(true)
+                    .setPreferPomsFromWorkspace(true))
+                    .getCurrentProject();
```

Comment:
I just realized there is `BootstrapWorkspaceProvider workspaceProvider;` injected in this Mojo, which caches the loaded workspace instead of reloading for every extension in the project which will be lot in projects such as Quarkus and Camel Quarkus. It's there exactly for this reason.
You could obtain a local project with `workspaceProvider.getProject(groupId, artifactId)`. Although it's not based on the effective models and may return null in some cases. Would you mind refactoring to use the `workspaceProvider` instead? We could look into initializing it differently in a separate PR.

Reply:
Will do. 

Update: Have tried, but it's not working well. With the following code

```
        final Artifact artifact = project.getArtifact();
        LocalProject localProject = workspaceProvider.getProject(artifact.getGroupId(), artifact.getArtifactId());
```
... localProject is always null. I can put that in and then do a null check and then go down the original path, which would be a performance optimisation in cases where `localProject` isn't null. But I can't tell how many of those there would be - is it just an unlucky setup on my test project, or is it likely to be null in many cases. And trying it both ways means more code to maintain. 
Did I invoke it wrong?

Reply:
Could I try it on your test project?

Reply:
That would be great. I'm just using the ones in the `resources` directory of the PR, such as https://github.com/quarkusio/quarkus/blob/872202d71b3c2596cbb8f3441786687bca18e045/independent-projects/extension-maven-plugin/src/test/resources/simple-pom-with-checks-disabled/pom.xml. I've pushed what I have, but it actually doesn't even need my branch to see problems, just the following in `addSource()` shows the null: 

```
       final Artifact artifact = project.getArtifact();
        System.out.println("HOLLY looking for " + project.getArtifact());
        LocalProject localProject = workspaceProvider.getProject(artifact.getGroupId(), artifact.getArtifactId());
        System.out.println("HOLLY local project is " + localProject);
```

Reply:
It's probably because it's using `Paths.get("")` to load the workspace instead of the test project dir.

Reply:
Ah, yes, I tried it by just building it and then invoking it the 'normal' way on an independent project, and it found the `<scm>` entry and put it into the yaml. It also works where the `<scm>` is in a local parent, and correctly ignores remote parents.

That's a bit awkward - it works, as long as I don't try and test it. :)

Reply:
So I guess I can 
- do it the fast way, with a fallback to the slow way, to keep the tests happy; it should be most robust, but the code is annoyingly complex, and the tests aren't actually testing the 'real' path which makes them a bit pointless
- do it the fast way and drop the tests (but that makes me nerrrrvous))
- find something I can do in the tests to make the workspace loading stick closer to the 'real' behaviour?

Reply:
It looks like if you add a new constructor to the `BootstrapWorkspaceProvider` that accepts a `Path` to the target project, you could provide then an instance of it initialized the way you need in a test in `makeMojo` in similar it is done for the `repoSystem` and `repoSession`.

Reply:
Or you could make the new constructor private an add a static method such as `newInstance(Path)` to not expose a possibly new option to the DI mechanism initializing mojos.

Reply:
Thanks for the pointer, @aloubyansky. That was exactly the issue, and I think the fix is a good one, since it makes the behaviour of the whole mojo a bit closer to the intention of the tests. It may help other tests in future, too. 

I have pushed and optimistically squashed. :)